zrml/swaps/src/fixed.rs

```diff
@@ -180,14 +232,158 @@ mod tests {
     fn bpow_has_minimum_set_of_correct_values() {
-        create_tests!(
-            bpow;
-            0 => Ok(BASE), Ok(0), Ok(0), Ok(0);
-            1 => Ok(BASE), Ok(BASE), Ok(BASE), Ok(BASE);
-            2 => Ok(BASE), Ok(2 * BASE), Ok(4 * BASE), Ok(8 * BASE);
-            3 => Ok(BASE), Ok(3 * BASE), Ok(9 * BASE), Ok(27 * BASE);
-            max_n => Ok(BASE), Ok(u128::MAX), ERR, ERR;
-            n_max => ERR, ERR, ERR, ERR;
-        );
+        let test_vector: Vec<(u128, u128, u128)> = vec![
+            (2500000000, 0, 10000000000),
+            (2500000000, 10000000000, 2500000000),
+            (2500000000, 33333333333, 98431332),
+            (2500000000, 200000000, 9726549474),
+            (2500000000, 500000000000, 0),
+            (5000000000, 0, 10000000000),
+            (5000000000, 10000000000, 5000000000),
+            (5000000000, 33333333333, 992125657),
+            (5000000000, 200000000, 9862327044),
+            (5000000000, 500000000000, 0),
+            (7500000000, 0, 10000000000),
+            (7500000000, 10000000000, 7500000000),
+            (7500000000, 33333333333, 3832988750),
+            (7500000000, 200000000, 9942628790),
+            (7500000000, 500000000000, 5663),
+            (10000000000, 0, 10000000000),
+            (10000000000, 10000000000, 10000000000),
+            (10000000000, 33333333333, 10000000000),
+            (10000000000, 200000000, 10000000000),
+            (10000000000, 500000000000, 10000000000),
+            (12500000000, 0, 10000000000),
+            (12500000000, 10000000000, 12500000000),
+            (12500000000, 33333333333, 21039401269),
+            (12500000000, 200000000, 10044728444),
+            (12500000000, 500000000000, 700649232162408),
+            (15000000000, 0, 10000000000),
+            (15000000000, 10000000000, 15000000000),
+            (15000000000, 33333333333, 38634105686),
+            (15000000000, 200000000, 10081422716),
+            (15000000000, 500000000000, 6376215002140495869),
+            (17500000000, 0, 10000000000),
+            (17500000000, 10000000000, 17500000000),
+            (17500000000, 33333333333, 64584280985),
+            (17500000000, 200000000, 10112551840),
+            (17500000000, 500000000000, 14187387615511831479362),
+        ];
+        for (base, exp, expected) in test_vector.iter() {
+            let result = bpow(*base, *exp).unwrap();
+            let precision = *expected / BASE + 4 * BPOW_PRECISION; // relative + absolute error
+            let diff = if result > *expected { result - *expected } else { *expected - result };
+            assert_le!(diff, precision);
+        }
+    }
+
+    #[test]
+    fn bpow_returns_error_when_parameters_are_outside_of_specified_limits() {
+        let test_vector: Vec<(u128, u128)> =
+            vec![(BASE / 10, 3 * BASE / 2), (2 * BASE - BASE / 10, 3 * BASE / 2)];
+        for (base, exp) in test_vector.iter() {
+            assert!(bpow(*base, *exp).is_err());
```

Comment:
I think this approach of testing the result is sufficient, but a more pedantic test could incorporate the exact errors that are excepted. I think in that case it is not necessary, but generally some other change might introduce another error in one of the test cases that is thrown before the excepted errors.
The check could look like this:

```rust
let errors = vec![
    Err(DispatchError::Other("[bpow_approx]: expected base >= BASE / 4")),
    Err(DispatchError::Other("[bpow_approx]: expected base <= 7 * BASE / 4"))
];
let test_vector: Vec<(u128, u128)> =
    vec![(BASE / 10, 3 * BASE / 2), (2 * BASE - BASE / 10, 3 * BASE / 2)];
for (base, exp) in test_vector.iter() {
    assert!(errors.contains(&bpow(*base, *exp)));
}
```

An even cleaner approach would be to open up two different test cases for the respective expected errors.